Realm/RLMRealm.mm

```diff
@@ -196,16 +187,26 @@ static void clearRealmCache() {
         for (NSMapTable *map in s_realmsPerPath.allValues) {
             [map removeAllObjects];
         }
-        s_realmsPerPath = [NSMutableDictionary dictionary];
+        [s_realmsPerPath removeAllObjects];
     }
 }
 
+static NSMutableDictionary *s_listenersPerPath = [NSMutableDictionary new];
+// must be called with s_listenersPerPath locked
+static RLMChangeListener *listenerForPath(NSString *path, BOOL inMemory, BOOL create=YES) {
+    RLMChangeListener *listener = s_listenersPerPath[path];
```

Comment:
synchronization should be pushed from the caller into this and other listener api methods